testing/src/test/java/io/stargate/it/driver/CqlSessionExtension.java

```diff
@@ -0,0 +1,405 @@
+ *
+ * Certain aspects of the session creation can be controlled with the {@link CqlSessionSpec}
+ * annotation. See the javadocs of that type for more details.
+ *
+ * <h3>Parameter injection</h3>
+ *
+ * The extension will detect and inject the following parameters on JUnit test methods
+ * (constructors, BeforeXxx or test methods, etc):
+ *
+ * <ul>
+ *   <li>any parameter of type {@link CqlSession}: injected with the session that was created by the
+ *       extension.
+ *   <li>any parameter of type {@link CqlSessionBuilder}: injected with a new session builder
+ *       pre-configured to connect to the Stargate cluster. This allows you to create new sessions
+ *       with a different configuration in your test methods. Note that such sessions are not
+ *       managed by the extension: <b>you must close them explicitly once you are done</b>, to avoid
+ *       resource leaks. A try-with-resources block is generally the best way to do that. Also, you
+ *       might want to invoke {@link #awaitAllNodes(CqlSession, StargateEnvironmentInfo)} if there
+ *       are multiple Stargate nodes.
+ *   <li>a parameter of type {@link CqlIdentifier} <b>AND annotated with {@link TestKeyspace}</b>:
+ *       injected with the identifier of the keyspace that was created by the extension. This is
+ *       useful if you need to look up schema metadata, use the keyspace from another session or
+ *       API, etc.
+ * </ul>
+ */
+public class CqlSessionExtension
+    implements BeforeAllCallback,
+        BeforeEachCallback,
+        AfterEachCallback,
+        AfterAllCallback,
+        ParameterResolver {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CqlSessionExtension.class);
+  private static final int KEYSPACE_NAME_MAX_LENGTH = 48;
+
+  // Tracks whether the extension was configured per-class or per-method.
+  private volatile Lifecycle sessionLifecycle;
+
+  private volatile StargateEnvironmentInfo stargate;
+  private volatile CqlSessionSpec cqlSessionSpec;
+  private volatile CqlIdentifier keyspaceId;
+  private volatile CqlSession session;
+
+  @Override
+  public void beforeAll(ExtensionContext context) {
+    LOG.debug("Starting per-class execution for {}", context.getElement());
+    // beforeAll and afterAll are never called if we're per-method, that's how we figure out the
+    // lifecycle
+    sessionLifecycle = Lifecycle.PER_CLASS;
+    createSession(context);
+  }
+
+  @Override
+  public void beforeEach(ExtensionContext context) {
+    if (sessionLifecycle == null) {
+      LOG.debug("Starting per-method execution for {}", context.getElement());
+      sessionLifecycle = Lifecycle.PER_METHOD;
+      createSession(context);
+    }
+  }
+
+  @Override
+  public void afterEach(ExtensionContext context) {
+    assert sessionLifecycle != null;
+    if (sessionLifecycle == Lifecycle.PER_METHOD) {
+      LOG.debug("Stopping per-method execution for {}", context.getElement());
+      destroySession(context);
+    }
```

Comment:
If we store something implementing `ExtensionContext.Store.CloseableResource`, JUnit 5 will automatically close that thing when the context is closed. `ExternalResource` uses this approach. I do not insist, though :)

Reply:
I don't see a strong advantage either way, so I'll keep it like this, it's more explicit.